### PR TITLE
Add guide about C++ unit test.

### PR DESCRIPTION
**Pull request by bianpengyuan**
Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

### CODE REVIEW

**Comment by bianpengyuan** [comment]
Any help on reviewing this? Thanks!

**Comment by bianpengyuan** [comment]
Gentle ping. Can I please get some review on this? Thanks!

File: doc/write-cpp-unit-test.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+In [`proxy-wasm-cpp-host`](https://github.com/proxy-wasm/proxy-wasm-cpp-host), along with `V8`, `wasmtime`, `WAVM` runtimes, which could interpret Wasm binaries, [a `nullVM` runtime](https://github.com/proxy-wasm/proxy-wasm-cpp-host/tree/master/src/null) is also included, which could compile together with the extension code and run natively. The unit test will utilize `nullVm` runtime. Specifically, all host implementation of `proxy-wasm-cpp-host` (such as Envoy) needs to implements a [`Context` interface class](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/eceb02d5b7772ec1cd78a4d35356e57d2e6d59bb/include/proxy-wasm/context.h#L128), which has many unimplemented host specific methods. In the unit test, a mock implementation of the `Context` class will be created using [GoogleTest framework](https://github.com/google/googletest), and compiled with your extension code. Within the mock implementation, we could inject desired host behavior and check if the extension interacts with host as expected.
```
Comment (kyessenov) [inline doc/write-cpp-unit-test.md:10]: ... runtimes that execute Wasm binaries... ... which runs natively compiled-in modules.

File: doc/write-cpp-unit-test.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+#ifdef NULL_PLUGIN
+
+}  // namespace basic_auth
+}  // namespace null_plugin
+}  // namespace proxy_wasm
+
+#endif
+```
+
+Then to build the extension under `nullVM` mode, the following target is added to the `BUILD` file. The target is similiar to the one generates Wasm binary, except that it uses `cc_binary` and defines `NULL_PLUGIN` macro in `copts`.
```
Comment (kyessenov) [inline doc/write-cpp-unit-test.md:71]: _generated_

File: doc/write-cpp-unit-test.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    deps = [
+        "@com_google_absl//absl/strings",
+        "@com_google_absl//absl/time",
+        "@io_istio_proxy//extensions/common/wasm:json_util",
+        "@proxy_wasm_cpp_host//:lib",
+    ],
+)
+```
+
+## Step 2: Create scaffold for unit tests
```
Comment (kyessenov) [inline doc/write-cpp-unit-test.md:94]: _a scaffold_

File: doc/write-cpp-unit-test.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+---
+
+Before implementing test, the scaffold needs to be created, which imports necessary library, adds the namespace wrapper, registers the extension with `nullVM` runtime, and creates a test class which intiliaze the `nullVM` runtime. Unlike extension code, the unit test is always compiled with nullVM mode, so there is no need to wrap the namespace with `NULL_PLUGIN` directive.
```
Comment (kyessenov) [inline doc/write-cpp-unit-test.md:97]: _initializes_

**Review by kyessenov — APPROVED** [review]
Looks good, just some writing suggestions.

**Pull request merged**